File: 033_TimeBasedKeyValueStore.cpp
// Topic: Time Based Key-Value Store
// Description: Design structure supporting timestamped key-value retrievals.
// Approach: For each key keep sorted vector of (time, value) and binary search for floor timestamp.

#include <bits/stdc++.h>
using namespace std;

class TimeMap
{
    unordered_map<string, vector<pair<int, string>>> mp;

   public:
    TimeMap() { mp.reserve(1024); }
    void set(const string& key, const string& value, int timestamp)
    {
        mp[key].emplace_back(timestamp, value);
    }
    string get(const string& key, int timestamp)
    {
        auto it = mp.find(key);
        if (it == mp.end())
            return "";
        auto& v = it->second;
        int l = 0, r = (int)v.size() - 1, ans = -1;
        while (l <= r)
        {
            int mid = l + ((r - l) >> 1);
            if (v[mid].first <= timestamp)
            {
                ans = mid;
                l = mid + 1;
            }
            else
                r = mid - 1;
        }
        return ans == -1 ? string() : v[ans].second;
    }
};

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    TimeMap tm;
    tm.set("foo", "bar", 1);
    cout << "Test 1: get(foo,1) -> " << tm.get("foo", 1) << "\n";  // bar
    cout << "Test 1: get(foo,3) -> " << tm.get("foo", 3) << "\n";  // bar
    tm.set("foo", "bar2", 4);
    cout << "Test 1: get(foo,4) -> " << tm.get("foo", 4) << "\n";  // bar2
    cout << "Test 1: get(foo,5) -> " << tm.get("foo", 5) << "\n";  // bar2
}

/*
Time Complexity: set O(1) amortized, get O(log n_k) for key's list size
Space Complexity: O(total entries)
*/

File: 033_TimeBasedKeyValueStore.py
# Topic: Time Based Key-Value Store
# Description: Design structure supporting timestamped key-value retrievals.
# Approach: For each key keep sorted list of (time, value) and binary search for floor timestamp.

from collections import defaultdict
import bisect


class TimeMap:
    def __init__(self):
        self.store = defaultdict(list)

    def set(self, key: str, value: str, timestamp: int) -> None:
        self.store[key].append((timestamp, value))

    def get(self, key: str, timestamp: int) -> str:
        if key not in self.store:
            return ""
        arr = self.store[key]
        # Binary search for largest timestamp <= given timestamp
        idx = bisect.bisect_right(arr, (timestamp, chr(127))) - 1
        if idx < 0:
            return ""
        return arr[idx][1]


if __name__ == "__main__":
    tm = TimeMap()
    tm.set("foo", "bar", 1)
    print(f"Test 1: get(foo,1) -> {tm.get('foo', 1)}")  # bar
    print(f"Test 1: get(foo,3) -> {tm.get('foo', 3)}")  # bar
    tm.set("foo", "bar2", 4)
    print(f"Test 1: get(foo,4) -> {tm.get('foo', 4)}")  # bar2
    print(f"Test 1: get(foo,5) -> {tm.get('foo', 5)}")  # bar2

"""
Time Complexity: set O(1) amortized, get O(log n_k) for key's list size
Space Complexity: O(total entries)
"""
